### PR TITLE
7 speech to text for voice commands (local container)

### LINKED ISSUE #12: Facial recognition (Azure) to identify individuals rather than body features as in demos

### LINKED ISSUE #121: Face API integration

### PR DESCRIPTION
**Pull request by FergusKidd**
fixes #121 fixes #12

### CODE REVIEW

File: control-agent/text-to-speech/README.md [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+Recalibrate yourself
+
+# Set Up
+
+In order to set this up for yourseld you will need the following named intents in LUIS:
+
+![LUISintents](./docs/images/LUISintents.png)
+
+and an optional QnA.
+`tip: you can automatically flesh out a lot of questiosn by assigning one of the pre-built personalities to your robot`
```
Comment (Sealjay-clj) [inline control-agent/text-to-speech/README.md:22]:
```suggestion
`tip: you can automatically flesh out a lot of questions by assigning one of the pre-built personalities to your robot`
```

File: control-agent/text-to-speech/README.md [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+Move direction
+Close/Open Grip
+What can you see?
+What is the time?
+Put your arm up/down/in/out
+Recalibrate yourself
+
+# Set Up
+
+In order to set this up for yourseld you will need the following named intents in LUIS:
```
Comment (Sealjay-clj) [inline control-agent/text-to-speech/README.md:17]:
```suggestion
In order to set this up for yourself you will need the following named intents in LUIS:
```

File: control-agent/text-to-speech/README.md [hunk 3]
```diff
@@ -0,0 +1,22 @@
+To install this loval chatbot functionality simply move the contents of this folder onto the stretch, and run
```
Comment (Sealjay-clj) [inline control-agent/text-to-speech/README.md:1]:
```suggestion
To install this local chatbot functionality simply move the contents of this folder onto the stretch, and run
```

File: control-agent/text-to-speech/config-template.json [hunk 4]
```diff
@@ -10,6 +10,8 @@
     "LuisKey":"",
     "LuisRegion":"",
     "qnaURL":"",
-    "qnaKey":""
+    "qnaKey":"",
+    "FaceKey":"",
```
Comment (Sealjay-clj) [inline control-agent/text-to-speech/config-template.json:14]: Should probably standardise capitalisation in a future update

File: control-agent/text-to-speech/moveFB.py [hunk 5]
```diff
@@ -0,0 +1,20 @@
+import stretch_body.robot
+import time
```
Comment (Sealjay-clj) [inline control-agent/text-to-speech/moveFB.py:2]: Did you pylint this? Maybe check the import order

File: control-agent/text-to-speech/moveFB.py [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+import sys
+
+robot = stretch_body.robot.Robot()
+robot.startup()
+move_by = 0
+
+print("recieved:", sys.argv[1], sys.argv[2])
```
Comment (Sealjay-clj) [inline control-agent/text-to-speech/moveFB.py:9]: We've used fstrings elsewhere

File: control-agent/text-to-speech/movearm.py [hunk 7]
```diff
@@ -0,0 +1,26 @@
+import stretch_body.robot
+import sys
+import time
+
+robot = stretch_body.robot.Robot()
+robot.startup()
+
+
+if sys.argv[3] == "in":
```
Comment (Sealjay-clj) [inline control-agent/text-to-speech/movearm.py:9]: I'd move sys.argv[3] to a variable, and then check if the code is running on main to make it more reusable

File: control-agent/text-to-speech/realsense.py [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+import time
+import math
+
+from PIL import Image
+import io
+
+import speech
+
+
+def take_photo():
```
Comment (Sealjay-clj) [inline control-agent/text-to-speech/realsense.py:12]: Consider breaking this up into smaller functions

File: control-agent/text-to-speech/speech-detection.py [hunk 9]
```diff
@@ -132,26 +133,82 @@ def QnA(question):
...
+            if (
+                "person"
+                or "boy"
+                or "man"
+                or "woman" in result["description"]["captions"][0]["text"]
+            ):
+                # TODO more than one person?
+                try:
+                    person = speech.identify_face(
+                        speech.recognize_face(image)[0]["faceId"]
```
Comment (Sealjay-clj) [inline control-agent/text-to-speech/speech-detection.py:158]: If you break this out into a function, you could iterate over the  results of speech.recognize_face. E.g. for face in recognition_results: append_name

https://stackoverflow.com/questions/53981845/grammatically-correct-human-readable-string-from-list-with-oxford-comma

File: control-agent/text-to-speech/speech-detection.py [hunk 10]
```diff
@@ -132,26 +133,82 @@ def QnA(question):
 
 run = True
 
+speech.speak("starting up")
+
 while run == True:
 
     intent = recognize_intent()
     try:
```
Comment (Sealjay-clj) [inline control-agent/text-to-speech/speech-detection.py:138]: When Python 3.10 comes out, this is ripe for match case ;)

File: control-agent/text-to-speech/speech-detection.py [hunk 11]
```diff
@@ -162,11 +219,72 @@ def QnA(question):
             now = datetime.now()
             speakabletime = now.strftime("%H:%M")
             speech.speak("The time is, " + speakabletime)
+
         elif intent.intent_id == "Stop":
             break
+        elif intent.intent_id == "Selfie":
```
Comment (Sealjay-clj) [inline control-agent/text-to-speech/speech-detection.py:225]: In future, I'd pass each intent to a function. You could have a dictionary of function references

File: control-agent/text-to-speech/speech.py [hunk 12]
```diff
@@ -145,14 +159,99 @@ def recognize(blobData):
...
+
+    headers = {
+        # Request headers
+        "Content-Type": "application/octet-stream",
+        "Ocp-Apim-Subscription-Key": key,
+    }
+
+    params = urllib.parse.urlencode(
+        {
+            # Request parameters
```
Comment (Sealjay-clj) [inline control-agent/text-to-speech/speech.py:183]:
```suggestion
           
```

File: control-agent/text-to-speech/speech.py [hunk 13]
```diff
@@ -145,14 +159,99 @@ def recognize(blobData):
...
+
+    key, url = getFaceKeys()
+
+    headers = {
+        # Request headers
+        "Content-Type": "application/octet-stream",
+        "Ocp-Apim-Subscription-Key": key,
+    }
+
+    params = urllib.parse.urlencode(
```
Comment (Sealjay-clj) [inline control-agent/text-to-speech/speech.py:181]: If you rename this variable to request_params, you don't need the comment

File: control-agent/text-to-speech/speech.py [hunk 14]
```diff
@@ -145,14 +159,99 @@ def recognize(blobData):
...
+    except Exception as e:
+        print("Vision Error: ", e)
+        return "error"
+
+    # face logic
+    # TODO remove hard coding
+    results = json.loads(data)
+    if (
+        results[0]["candidates"][0]["personId"]
+        == "f9611b03-dc30-48bd-88f5-9ce251f688b6"
```
Comment (Sealjay-clj) [inline control-agent/text-to-speech/speech.py:248]: Put this in a variable with a single leading underscore whilst it's hard coded -  https://pep8.org/

File: control-agent/text-to-speech/speech.py [hunk 15]
```diff
@@ -145,14 +159,99 @@ def recognize(blobData):
...
+        conn.close()
+    except Exception as e:
+        print("Vision Error: ", e)
+        return "error"
+
+    # face logic
+    # TODO remove hard coding
+    results = json.loads(data)
+    if (
+        results[0]["candidates"][0]["personId"]
```
Comment (Sealjay-clj) [inline control-agent/text-to-speech/speech.py:247]: While it's hard coded, you could put this in a dictionary

**Review by Sealjay-clj — COMMENTED** [review]
Some minor stylistic changes, and potential improvements to refactor noted, albeit it looks like it works

File: control-agent/text-to-speech/speech-detection.py (hunk 11 above)
Comment (FergusKidd) [reply control-agent/text-to-speech/speech-detection.py:225]: Yes that's a good idea, I'll put in a card for a refactor

File: control-agent/text-to-speech/speech.py (hunk 15 above)
Comment (FergusKidd) [reply control-agent/text-to-speech/speech.py:247]: added as dict so more can be added later

File: control-agent/text-to-speech/speech-detection.py (hunk 9 above)
Comment (FergusKidd) [reply control-agent/text-to-speech/speech-detection.py:158]: I'll raise another issue

File: control-agent/text-to-speech/moveFB.py (hunk 5 above)
Comment (FergusKidd) [reply control-agent/text-to-speech/moveFB.py:1]: reordered

File: control-agent/text-to-speech/movearm.py (hunk 7 above)
Comment (FergusKidd) [reply control-agent/text-to-speech/movearm.py:10]: done for all sys args

File: control-agent/text-to-speech/moveFB.py (hunk 6 above)
Comment (FergusKidd) [reply control-agent/text-to-speech/moveFB.py:10]: This debug print can be removed anyway

File: control-agent/text-to-speech/realsense.py (hunk 8 above)
Comment (FergusKidd) [reply control-agent/text-to-speech/realsense.py:9]: adding to refactor issue

**Pull request merged**